### PR TITLE
front: sort curves in rs editor

### LINKED ISSUE #5612: Effort curves are not sorted properly when sent to back-end
### What happened?

Speed-Effort curves in the rolling stock editor are not sent in the right order to backend, and the default curve is sometimes not the right one.

### What did you expect to happen?

_This is a temporary fix awaiting for the data model and the rs editor interface to match better. This will be done later._

The order to be passed to the back end (sorting to be done by the front).

For a given traction mode :
- Curves with a not-null restriction code and electrical profile must be grouped at the beginning of the list
- then group the curves that have a not-null power restriction code
- then group the ones that have a not-null electrical profile
- then group the ones remaining (there can only be 3 left, for the three comfort levels)

This order is set this way in order to favour the parameters that have the strongest impact on running-time calculation.

The curve with power restriction AND electrical profile set to null and a standard comfort level must be the default curve for the traction mode, and removed from the list.

_NB : We need to make sure the comfort level is never null, in the existing rolling stocks and in the one that can be produced by the editor !_

### OSRD version (top right corner `Account` button > `Informations`)

4778daf5

### PR DESCRIPTION
**Pull request by SarahBellaha**
Closes #5612

### CODE REVIEW

File: front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorCurves.tsx [hunk 1]
```diff
@@ -339,10 +340,15 @@ export default function RollingStockEditorCurves({
...
-          : currentRsEffortCurve.modes[selectedTractionMode].default_curve;
+
+      const isDefaultCurve =
+        selectedComfortLvl === STANDARD_COMFORT_LEVEL &&
+        selectedElectricalProfile === null &&
+        selectedPowerRestriction === null;
+
+      const defaultCurve = isDefaultCurve
+        ? updatedSelectedCurve.curve
+        : currentRsEffortCurve.modes[selectedTractionMode].default_curve;
```
Comment (clarani) [inline front/src/modules/rollingStock/components/RollingStockEditor/RollingStockEditorCurves.tsx:351]: Nice catch :)

File: front/src/modules/rollingStock/helpers/utils.ts [hunk 2]
```diff
@@ -193,5 +198,26 @@ export const orderSelectorList = (list: (string | null)[]) => {
...
+      } else if (cond?.electrical_profile_level && !cond?.power_restriction_code) {
+        arrIndex = 2;
+      } else if (!cond?.electrical_profile_level && !cond?.power_restriction_code) {
+        arrIndex = 3;
+      }
+      sortedCurves[arrIndex].push(curve);
+      return sortedCurves;
+    },
+    [[], [], [], []] as ConditionalEffortCurve[][]
+  );
```
Comment (clarani) [inline front/src/modules/rollingStock/helpers/utils.ts]: Here you can use a object, I think it's more readable than an array (but the idea is the same). What do you think ?
```suggestion
  const { allConds, onlyPR, onlyEP, nulls } = curvesList.reduce(
    (sortedCurves, curve) => {
      const { cond } = curve;
      if (!cond) return sortedCurves;
      if (cond.electrical_profile_level && cond.power_restriction_code) {
        return { ...sortedCurves, allConds: [...sortedCurves.allConds, curve] };
      }
      if (cond.power_restriction_code) {
        return { ...sortedCurves, onlyPR: [...sortedCurves.allConds, curve] };
      }
      if (cond.electrical_profile_level) {
        return { ...sortedCurves, onlyEP: [...sortedCurves.allConds, curve] };
      }
      return { ...sortedCurves, nulls: [...sortedCurves.allConds, curve] };
    },
    { allConds: [], onlyPR: [], onlyEP: [], nulls: [] } as Record<string, ConditionalEffortCurve[]>
  );
```

File: front/src/modules/rollingStock/helpers/utils.ts [hunk 3]
```diff
@@ -193,5 +198,26 @@ export const orderSelectorList = (list: (string | null)[]) => {
     : list.sort();
 };
 
+export const sortSelectedModeCurves = (curvesList: ConditionalEffortCurve[]) => {
```
Comment (clarani) [inline front/src/modules/rollingStock/helpers/utils.ts:240]: Nice function ! Can you add a comment to explain how we sort the curves and some unit tests ? (unit tests will be quick to write, you can use empty curves and only change the conditions)

**Review by clarani — CHANGES_REQUESTED** [review]
Nice PR :)

File: front/src/modules/rollingStock/helpers/utils.ts (hunk 2 above)
Comment (SarahBellaha) [reply front/src/modules/rollingStock/helpers/utils.ts]: I agree :)

File: front/src/modules/rollingStock/components/RollingStockEditor/__tests__/RollingStockEditorCurves.spec.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
+import {
+  createEmptyCurve,
+  createEmptyCurves,
+  sortSelectedModeCurves,
+} from 'modules/rollingStock/helpers/utils';
+
+const emptyEffortCurves = createEmptyCurves('1500', ['STANDARD', 'AC', 'HEATING']);
+const newCurveWithAllCond = createEmptyCurve('STANDARD', 'A1', 'C1US');
+const newCurveWithOnlyPR = createEmptyCurve('STANDARD', null, 'C1US');
+const newCurveWithOnlyEP = createEmptyCurve('STANDARD', null, 'C1US');
```
Comment (clarani) [inline front/src/modules/rollingStock/components/RollingStockEditor/__tests__/RollingStockEditorCurves.spec.ts]:
```suggestion
const newCurveWithOnlyEP = createEmptyCurve('STANDARD', 'A1');
```

File: front/src/modules/rollingStock/components/RollingStockEditor/__tests__/RollingStockEditorCurves.spec.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      newCurveWithOnlyEP,
+      newCurveWithAllCond,
+    ]);
+    const expected = [newCurveWithAllCond, newCurveWithOnlyEP, ...emptyEffortCurves.curves];
+
+    expect(result).toEqual(expected);
+  });
+});
+
+// --------- alternative ----------
```
Comment (clarani) [inline front/src/modules/rollingStock/components/RollingStockEditor/__tests__/RollingStockEditorCurves.spec.ts]: The last test is very relevant too, you can remove the describe and move the test in the previous describe :)
```suggestion
```

File: front/src/modules/rollingStock/helpers/utils.ts [hunk 6]
```diff
@@ -198,22 +198,31 @@ export const orderSelectorList = (list: (string | null)[]) => {
     : list.sort();
 };
 
+/** This function will sort the curves list based on the number of conditions they have.  */
```
Comment (clarani) [inline front/src/modules/rollingStock/helpers/utils.ts]: You can create 1 unique comment, what do you think ?
```suggestion
/**
  * This function will sort the curves list based on the number of conditions they have.
  *
  * Curves will be sorted in the following order (comfort is compulsory): 
  * - curves with all conditions (power restriction & electrical profile)
  * - curves with at least power restriction
  * - curves with least electrical profile
  * - other curves (without power restriction and without electrical profile)
  */
```

**Review by clarani — APPROVED** [review]
LGTM :white_check_mark:
